In [3]:
import pandas as pd

dir_suffix = "data/"
dir_suffix_entpackt = dir_suffix + "Daten_ausgepackt/"

pd.read_csv(dir_suffix + "tip25W_testdaten_template.csv")
df_prio = pd.read_csv(dir_suffix_entpackt + "order_products__prior.csv")
df_train = pd.read_csv(dir_suffix_entpackt + "order_products__train.csv")
df_products = pd.read_csv(dir_suffix_entpackt+"products.csv")

df_all = pd.concat([df_prio, df_train])

df_departments = pd.read_csv(dir_suffix_entpackt + "departments.csv")
df_departments = df_departments[df_departments["department_id"].between(1,5)]
df_departments

department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

Aufgabe 1:
*Berechnen Sie die 5x5 Matrix M, die an Position m i,k jeweils angibt wie wahrscheinlich es ist, dass die
nachfolgende Bestellung eines Users mindestens einen Artikel aus Department k enthält
(, unter der Bedingung, dass die vorherige Bestellung desselben Users mindestens einen Artikel aus
dem Department mit ID i enthalten hat. (
Um die Matrix überschaubar zu halten, beschränkt sich die Aufgabe auf die
Departments 1 bis 5. Die wahren Wahrscheinlichkeiten sind natürlich
unbekannt; sie müssen also nur einen sinnvollen Schätzwert berechnen. Bei
Matrizen wird der erste Index nach unten, der zweite nach rechts
dargestellt.*

Also:
Wenn ein User etwas aus Department i gekauft hat: Wie wahrscheinlich ist es dann, dass er in seiner nächsten Bestellung etwas aus Department k kauft?

m i,k = P(nächste Bestellung enthält Department k | vorherige Bestellung enthielt Department i)

diese Bestellung aus k ∩ letze Bestellung aus k
_______________________________________________
letzte Bestellung aus k


Beispiel:

Für i = 3 und k = 1

Für alle user die in Prior einen Artikel aus 3 hatten und in der jetzigen Bestellung etwas aus 1??

P(A) = jemand hat aus 3 bestellt 
P(B) = jemand hat aus 1 bestellt

P(A ∩ B) = jemand hat prior aus 3 und jetzt aus 1 bestellt


| i, k    | 1 | 2 | 3 | 4 | 5 |
| -------- | ------- | ------- | ------- | ------- | ------- |
| 1        |   |   |   |   |   |
| 2        |   |   |   |   |   |
| 3        |   |   |   |   |   |
| 4        |   |   |   |   |   |
| 5        |   |   |   |   |   |

In [4]:
import pandas as pd
overview = pd.merge(df_all[:100][['order_id', 'product_id']], df_products[['product_id', 'department_id']], on="product_id", how="left")
overview.head(5)

order_id  product_id  department_id
0         2       33120             16
1         2       28985              4
2         2        9327             13
3         2       45918             13
4         2       30035             13

In [5]:
from IPython.display import display

print("DF order Prio: ")
display(df_prio.head(1))

print("DF order train: ")
display(df_train.head(1))

print("DF orders:")
df_orders = pd.read_csv(dir_suffix_entpackt+"orders.csv")
display(df_orders.head(11))

print("DF products:")
display(df_products.head(1))

print("Departments:")
df_dep = pd.read_csv(dir_suffix_entpackt+"departments.csv")
display(df_dep.head(1))

DF order Prio: 


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1

DF order train: 


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1

DF orders:


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   1187899        1    train            11          4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

DF products:


product_id                product_name  aisle_id  department_id
0           1  Chocolate Sandwich Cookies        61             19

Departments:


department_id department
0              1     frozen

In [6]:
# Sort so latest orders are first per user
df_sorted = df_orders.sort_values(['user_id', 'order_number'], ascending=[True, True])
df_sorted.head(30)
#latest_two_per_user = df_sorted.groupby('user_id').head(2).reset_index(drop=True)
#latest_two_per_user.head(10)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   1187899        1    train            11          4                  8   
11   2168274        2    prior             1          2                 11   
12   1501582        2    prior             2          5                 10   
13   1901567        2    prior             3          1                 10   
14    738281        2    prior             4          2                 10   
15   1673511        2    prior             5          3                 11   
16   1199898        2    prior             6          2                  9   
17   3194192        2    prior             7          2                 12   
18    788338        2    prior             8          1                 15   
19   1718559        2    prior             9          2                  9   
20   1447487        2    prior            10          1                 11   
21   1402090        2    prior            11          1                 10   
22   3186735        2    prior            12          1                  9   
23   3268552        2    prior            13          4                 11   
24    839880        2    prior            14          3                 10   
25   1492625        2    train            15          1                 11   
26   1374495        3    prior             1          1                 14   
27    444309        3    prior             2          3                 19   
28   3002854        3    prior             3          3                 16   
29   2037211        3    prior             4          2                 18   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0  
26                     NaN  
27                     9.0  
28                    21.0  
29                    20.0

In [7]:
df_sorted = df_sorted[["user_id", "order_id", "eval_set", "order_number"]]


df_all_small = df_all[["product_id", "order_id"]]
df_all_small

product_id  order_id
0             33120         2
1             28985         2
2              9327         2
3             45918         2
4             30035         2
...             ...       ...
1384612       14233   3421063
1384613       35548   3421063
1384614       35951   3421070
1384615       16953   3421070
1384616        4724   3421070

[33819106 rows x 2 columns]

In [8]:
orders_merged = pd.merge(df_sorted, df_all_small, on="order_id",how="left")
orders_merged.head(30)

department_resolved = pd.merge(orders_merged[["user_id","product_id","eval_set","order_number"]],df_products[["department_id", "product_id"]], on="product_id", how="left")

department_resolved.drop(labels=["product_id"], inplace=True, axis=1)
department_resolved = department_resolved[department_resolved["department_id"].between(1,5)]
department_resolved = department_resolved.reset_index(drop=1)
department_resolved = department_resolved.sort_values(['user_id', 'order_number'], ascending=[True, True]).drop_duplicates().reset_index(drop=True)
department_resolved.head(30)


user_id eval_set  order_number  department_id
0         1    prior             2            4.0
1         1    prior             5            4.0
2         2    prior             1            4.0
3         2    prior             2            4.0
4         2    prior             3            4.0
5         2    prior             4            4.0
6         2    prior             4            1.0
7         2    prior             5            4.0
8         2    prior             6            1.0
9         2    prior             6            4.0
10        2    prior             7            4.0
11        2    prior             8            4.0
12        2    prior             8            1.0
13        2    prior             9            4.0
14        2    prior             9            1.0
15        2    prior            10            4.0
16        2    prior            10            1.0
17        2    prior            11            4.0
18        2    prior            11            1.0
19        2    prior            12            4.0
20        2    prior            12            3.0
21        2    prior            13            4.0
22        2    prior            13            3.0
23        2    prior            14            4.0
24        2    prior            14            1.0
25        2    train            15            1.0
26        2    train            15            4.0
27        3    prior             1            4.0
28        3    prior             2            4.0
29        3    prior             3            4.0

In [9]:
prev = (
    df[department_resolved['order_type'] == 'prior']
    [['user_id', 'department_id']]
    .drop_duplicates()
    .rename(columns={'department_id': 'prev_dept'})
)

latest = (
    df[df['order_type'] == 'latest']
    [['user_id', 'department_id']]
    .drop_duplicates()
    .rename(columns={'department_id': 'latest_dept'})
)


NameError: name 'df' is not defined

1. Bauen eines gesammten Datensatzes mit prior und train order_products_all

2. Produkte mit Department verknüpfen
    für jede order_id müssen wir wissen welche Products aus welchen Departments wurden bestellt.
    order_products_all mit products auf product_id joinen → jede Zeile bekommt eine deparment_id

3. Jetzt filtern wie alle orders raus die nicht aus Department 1 - 5 bestellt haben

4. jetzt haben wir eine Tabelle mit order_id, product_id, department_id (1 - 5)

5. jetzt wollen wir pro oder wissen aus welchem Department hat die jeweilige Order bestellt (mindestens eine aus 1 bis 5).
    Gruppieren nach order_id das heißt jetzt haben wir pro order die gekauften Department zuweisungen.
    ACHTUNG JETZT UNSERE ÜBERLEGUNG ... -> hier müssen wir alle duplikate rauswerfen da wir sonst mehrere Einträge pro Bestellung haben. Mehrere Produkte aus Department z.b.3 in derselben Bestellung zählen nur einmal → es geht nur darum, ob mindestens ein Artikel aus Department k drin war.

so ungefähr 

order_id	departments_in_order
10	        {2, 3}
11	        {1}
12	        {3, 5}


6. jetzt müssen wir wieder die Bestellungen mit dem User mergen da wir ja das Pro User machen müssen. 
    also mergen wir das oben nennen wir es mal order_depart mit orders (aber wir müssen alle train ignorieren) mit der order_id. Dazu nehmen wir noch das eval_set rein damit wir wissen ob es prior oder test ist.


user_id	order_number	order_id    eval_set	departments_in_order
1	    1	            2539329	    prior       {2, 4}
1	    2	            2398795	    prior       {3}
1	    3	            473747	    train        {1, 3, 5}







Formel pro intersection 




Anzahl aller Übergänge bei denen prior i war und next k
_________________________________________________________

Anzahl aller Übergänge bei denen prior k war



Beispiel



Anzahl aller Übergänge bei denen prior 3 war und next 1
_________________________________________________________

Anzahl aller Übergänge bei denen prior 3 war



wichtig ist wir dürfen für den nenner nicht nur prior und dep 3 zählen da man sonst auch alle zählt die keine nachfolgende bestellung haben es kann ja sein dass es gitb 



Da ist die Wahrscheinlichkeit dass das Übergang von i zu k passiert ist




In [36]:
#department_resolved.sort_values(['user_id', 'order_number'])

new_df = (
    department_resolved
    .groupby(["user_id", "order_number"])["department_id"]
    .agg(list)
    .reset_index()
)
new_df.head(30)

    

new_df['next_departments'] = ( new_df.groupby('user_id')['department_id'].shift(-1) )
new_df

transition = new_df.dropna(subset=["next_departments"])
transition.head(10)



user_id  order_number department_id next_departments
0         1             2         [4.0]            [4.0]
2         2             1         [4.0]            [4.0]
3         2             2         [4.0]            [4.0]
4         2             3         [4.0]       [4.0, 1.0]
5         2             4    [4.0, 1.0]            [4.0]
6         2             5         [4.0]       [1.0, 4.0]
7         2             6    [1.0, 4.0]            [4.0]
8         2             7         [4.0]       [4.0, 1.0]
9         2             8    [4.0, 1.0]       [4.0, 1.0]
10        2             9    [4.0, 1.0]       [4.0, 1.0]

In [ ]:
import numpy as np
transition_small = transition[:5]

erg_arr = [[]]
erg_arr = np.zeros((5,5), dtype=int)

for keys, group in transition.groupby(["user_id", "order_number"]):
    print(keys)
    
    for department_last in group["department_id"]:
        for department_value in department_last:
            
            #print("von: ", department_value,"->")
              
            for depart_next in group["next_departments"]:
                for val in depart_next:
                    #print("nach: ",val , "<-")
                    erg_arr[int(val)-1][int(department_value)-1] += 1
                     
    
     
   #print("----------------- \n")
    
erg_arr

(np.int64(1), np.int64(2))
von:  4.0 ->
nach:  4.0 <-
----------------- 

(np.int64(2), np.int64(1))
von:  4.0 ->
nach:  4.0 <-
----------------- 

(np.int64(2), np.int64(2))
von:  4.0 ->
nach:  4.0 <-
----------------- 

(np.int64(2), np.int64(3))
von:  4.0 ->
nach:  4.0 <-
nach:  1.0 <-
----------------- 

(np.int64(2), np.int64(4))
von:  4.0 ->
nach:  4.0 <-
von:  1.0 ->
nach:  4.0 <-
----------------- 



array([[0, 0, 0, 1, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [1, 0, 0, 5, 0],
       [0, 0, 0, 0, 0]])

In [32]:
import numpy as np

# 5x5 Matrix für Übergangszählungen
M_counts = np.zeros((5,5), dtype=int)

# Zähler wie oft ein Department als "prev" vorkommt
M_prev = np.zeros(5, dtype=int)

# Iteriere über alle Übergänge
for prev, nxt in zip(transition['department_id'], transition['next_departments']):
    
    # prev und nxt sind Listen, z.B. [4.0, 1.0]
    
    for i in prev:                     # für jedes Department im prev
        i_idx = int(i) - 1
        M_prev[i_idx] += 1             # prev-Zähler erhöhen
        
        for k in nxt:                  # für jedes Department im next
            k_idx = int(k) - 1
            M_counts[i_idx, k_idx] += 1
            

# Wahrscheinlichkeitsmatrix berechnen
M = M_counts / M_prev[:, None]

# Ausgabe
print("Übergangszählungen:\n", M_counts)
print("\nAnzahl prev pro Department:\n", M_prev)
print("\nÜbergangswahrscheinlichkeiten (5x5 Matrix):\n", M)


Übergangszählungen:
 [[ 725161   15822  410808  965267   27844]
 [  15490    7382   11631   26900    2621]
 [ 411640   11854  469308  735012   22796]
 [ 965415   27173  734373 2158351   44603]
 [  28959    2641   23572   46446   55016]]

Anzahl prev pro Department:
 [1143632   33500  854971 2335702   81792]

Übergangswahrscheinlichkeiten (5x5 Matrix):
 [[0.63408596 0.01383487 0.35921345 0.84403637 0.02434699]
 [0.46238806 0.22035821 0.34719403 0.80298507 0.07823881]
 [0.48146662 0.0138648  0.54891686 0.85969232 0.02666289]
 [0.4133297  0.01163376 0.31441211 0.92406951 0.01909619]
 [0.35405663 0.03228922 0.28819444 0.56785505 0.67263302]]


In [37]:
transition.head(20)

user_id  order_number department_id next_departments
0         1             2         [4.0]            [4.0]
2         2             1         [4.0]            [4.0]
3         2             2         [4.0]            [4.0]
4         2             3         [4.0]       [4.0, 1.0]
5         2             4    [4.0, 1.0]            [4.0]
6         2             5         [4.0]       [1.0, 4.0]
7         2             6    [1.0, 4.0]            [4.0]
8         2             7         [4.0]       [4.0, 1.0]
9         2             8    [4.0, 1.0]       [4.0, 1.0]
10        2             9    [4.0, 1.0]       [4.0, 1.0]
11        2            10    [4.0, 1.0]       [4.0, 1.0]
12        2            11    [4.0, 1.0]       [4.0, 3.0]
13        2            12    [4.0, 3.0]       [4.0, 3.0]
14        2            13    [4.0, 3.0]       [4.0, 1.0]
15        2            14    [4.0, 1.0]       [1.0, 4.0]
17        3             1         [4.0]            [4.0]
18        3             2         [4.0]            [4.0]
19        3             3         [4.0]       [4.0, 1.0]
20        3             4    [4.0, 1.0]       [4.0, 1.0]
21        3             5    [4.0, 1.0]            [4.0]